azure/src/test/java/org/apache/iceberg/azure/adlsv2/VendedAzureSasCredentialProviderTest.java

```diff
@@ -0,0 +1,323 @@
+    Credential credential =
+        ImmutableCredential.builder()
+            .prefix(CREDENTIAL_PREFIX)
+            .config(
+                ImmutableMap.of(
+                    ADLS_SAS_TOKEN_PREFIX + STORAGE_ACCOUNT,
+                    "randomSasToken",
+                    ADLS_SAS_TOKEN_EXPIRE_AT_MS_PREFIX + STORAGE_ACCOUNT,
+                    Long.toString(Instant.now().plus(1, ChronoUnit.HOURS).toEpochMilli())))
+            .build();
+    LoadCredentialsResponse response =
+        ImmutableLoadCredentialsResponse.builder().addCredentials(credential).build();
+    HttpResponse mockResponse =
+        response(LoadCredentialsResponseParser.toJson(response)).withStatusCode(200);
+    mockServer.when(mockRequest).respond(mockResponse);
+
+    try (VendedAzureSasCredentialProvider provider =
+        new VendedAzureSasCredentialProvider(
+            ImmutableMap.of(VendedAzureSasCredentialProvider.URI, URI))) {
+      AzureSasCredential azureSasCredential = provider.credentialForAccount(STORAGE_ACCOUNT);
+      assertThat(azureSasCredential.getSignature())
+          .isEqualTo(credential.config().get(ADLS_SAS_TOKEN_PREFIX + STORAGE_ACCOUNT));
+
+      for (int i = 0; i < 5; i++) {
+        // resolving credentials multiple times should not hit the credentials endpoint again
+        assertThat(provider.credentialForAccount(STORAGE_ACCOUNT)).isSameAs(azureSasCredential);
+      }
+    }
+    mockServer.verify(mockRequest, VerificationTimes.once());
+  }
+
+  @Test
+  public void expiredSasToken() throws InterruptedException {
+    HttpRequest mockRequest = request("/v1/credentials").withMethod(HttpMethod.GET.name());
+    Credential credential =
+        ImmutableCredential.builder()
+            .prefix(CREDENTIAL_PREFIX)
+            .config(
+                ImmutableMap.of(
+                    ADLS_SAS_TOKEN_PREFIX + STORAGE_ACCOUNT,
+                    "randomSasToken",
+                    ADLS_SAS_TOKEN_EXPIRE_AT_MS_PREFIX + STORAGE_ACCOUNT,
+                    Long.toString(Instant.now().minus(1, ChronoUnit.MINUTES).toEpochMilli())))
+            .build();
+    LoadCredentialsResponse response =
+        ImmutableLoadCredentialsResponse.builder().addCredentials(credential).build();
+    HttpResponse mockResponse =
+        response(LoadCredentialsResponseParser.toJson(response)).withStatusCode(200);
+    mockServer.when(mockRequest).respond(mockResponse);
+
+    try (VendedAzureSasCredentialProvider provider =
+        new VendedAzureSasCredentialProvider(
+            ImmutableMap.of(VendedAzureSasCredentialProvider.URI, URI))) {
+      AzureSasCredential azureSasCredential = provider.credentialForAccount(STORAGE_ACCOUNT);
+      assertThat(azureSasCredential.getSignature())
+          .isEqualTo(credential.config().get(ADLS_SAS_TOKEN_PREFIX + STORAGE_ACCOUNT));
+
+      Thread.sleep(20);
+      // Since expiration time past to current time, the refresh will fall back at minimum 10ms
```

Comment:
I'm not sure what this comment is trying to say. Also what happens if you remove the sleep time above?

Reply:
- Since the credential refresh implementation is based on scheduling a new refresh using the server provided expiration time, 
- this test is covering the scenario where server responds with already expired token (expires-at-ms < current-time). 
- The scheduling delay for consequent refresh is minimum delay of 10ms as per the logic in [refreshDelayMillis()](https://github.com/apache/iceberg/pull/11577/files#diff-6c5ab073d2ee71c62b8ff3fb717fe65d00f499043c03d036ce8bf29796cf31faR69)
- If we remove the sleep time, the credentials will be fetched from API only once. Depending on the test case execution time (> 10ms) there can be second credential fetch. To keep the test case behaviour consistent I used thread sleep.